src/lib.rs

```diff
@@ -248,31 +258,307 @@ impl<T, const N: usize> ArrayVec<T, { N }> {
         let len = self.len();
 
         // bounds checks
-        if index > self.len() {
+        if index > len {
             out_of_bounds!("try_insert", index, len);
         }
         if self.is_full() {
             return Err(CapacityError(item));
         }
 
         unsafe {
-            // The spot to put the new value
-            let p = self.as_mut_ptr().add(index);
-            // Shift everything over to make space. (Duplicating the
-            // `index`th element into two consecutive places.)
-            ptr::copy(p, p.offset(1), len - index);
-            // Write it in, overwriting the first copy of the `index`th
-            // element.
-            ptr::write(p, item);
-            // update the length
-            self.set_len(len + 1);
+            self.insert_unchecked(index, item, len);
         }
 
         Ok(())
     }
 
+    /// Insert an item into the vector, removing and returning its last
+    /// item if already full.
+    ///
+    /// # Panics
+    ///
+    /// The item cannot be inserted at an index greater than the
+    /// vector's length or greater or equal to the maximum capacity `N`.
+    ///
+    /// # Exmaples
+    ///
+    /// When the vector's not full, [`ArrayVec::force_insert`] acts like
+    /// [`ArrayVec::insert`]:
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector: ArrayVec<u8, 5> = ArrayVec::new();
+    ///
+    /// assert_eq!(vector.force_insert(0, 42), None);
+    /// assert_eq!(vector.force_insert(0, 24), None);
+    /// assert_eq!(&vector, [24, 42].as_ref());
```

Comment:
We need `as_ref()`: `slice::as_ref` explicitly returns a slice which makes `rustc` able to automatically understand that `&vector` should be a slice. Doing the same test with `&[...]` means  "compare a reference to an `ArrayVec` with a reference to an array" and, since there is no automatic `Deref` to turn one type into the other, `rustc` just gives up with an error saying that they can't be compared (at least that's how I understood it).